File: AjaxVectorObjects/Math/RotatedRectangleF.cs
// Copyright (c) 2018 Aurigma Inc. All rights reserved.
// Licensed under the MIT License. See LICENSE file in the project root for full license information.
//
using System.ComponentModel;
using System.Drawing;
using SMath = System.Math;

namespace Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math
{
    public class RotatedRectangleF
    {
        public RotatedRectangleF()
        {
            CenterX = 0;
            CenterY = 0;
            Width = 2;
            Height = 2;
            Angle = 0;
        }

        public RotatedRectangleF(float centerX, float centerY, float width, float height, float angle)
        {
            CenterX = centerX;
            CenterY = centerY;
            Width = width;
            Height = height;
            Angle = angle;
        }

        public RotatedRectangleF(RectangleF rect, float angle = 0)
            : this(rect.X + rect.Width / 2, rect.Y + rect.Height / 2, rect.Width, rect.Height, angle)
        {
        }

        public RectangleF Bounds
        {
            get
            {
                using (var path = GetPath().ToGdiPlusGraphicsPath())
                {
                    return path.GetBounds();
                }
            }
        }

        public AdvancedDrawing.Path GetPath()
        {
            var path = new AdvancedDrawing.Path();
            path.DrawRectangle(ToRectangleF());

            if (Angle != 0)
            {
                using (var m = new System.Drawing.Drawing2D.Matrix())
                {
                    m.RotateAt(Angle, Center.ToPointF());
                    path.ApplyTransform(m);
                }
            }

            return path;
        }

        public PointF Center
        {
            get { return new PointF(CenterX, CenterY); }

            set
            {
                CenterX = value.X;
                CenterY = value.Y;
            }
        }

        [DefaultValue("0")]
        public float CenterX { get; set; }

        [DefaultValue("0")]
        public float CenterY { get; set; }

        [DefaultValue("2")]
        public float Width { get; set; }

        [DefaultValue("2")]
        public float Height { get; set; }

        /// <summary>
        /// Angle of rotation in degrees.
        /// </summary>
        [DefaultValue("0")]
        public float Angle { get; set; }

        /// <summary>
        /// Get or set the coordinates of upper-left corner of this rectangle
        /// </summary>
        public PointF Location
        {
            get
            {
                return GetLeftTopEdge();
            }
            set
            {
                var p = Location;
                CenterX = CenterX + (value.X - p.X);
                CenterY = CenterY + (value.Y - p.Y);
            }
        }

        public void RotateAt(double angle, PointF center)
        {
            Center = Center.RotateAt(angle, center);
            Angle = (float)angle;
        }

        public void Scale(float scaleX, float scaleY)
        {
            Center = Center.Scale(scaleX, scaleY);
            Width *= scaleX;
            Height *= scaleY;
        }

        public static bool AreEqual(RotatedRectangleF rect1, RotatedRectangleF rect2, double tolerance = 0)
        {
            return Utils.EqualsOfFloatNumbers(rect1.CenterX, rect2.CenterX, tolerance) &&
                   Utils.EqualsOfFloatNumbers(rect1.CenterY, rect2.CenterY, tolerance) &&
                   Utils.EqualsOfFloatNumbers(rect1.Width, rect2.Width, tolerance) &&
                   Utils.EqualsOfFloatNumbers(rect1.Height, rect2.Height, tolerance) &&
                   Utils.EqualsOfFloatNumbers(rect1.Angle, rect2.Angle, tolerance);
        }

        public override bool Equals(object obj)
        {
            var equals = object.ReferenceEquals(this, obj) ||
                object.ReferenceEquals(obj, null) && object.ReferenceEquals(this, null);

            if (!equals)
            {
                var rect = obj as RotatedRectangleF;
                equals = rect != null && AreEqual(this, rect);
            }
            return equals;
        }

        public override int GetHashCode()
        {
            unchecked // Overflow is fine
            {
                // http://stackoverflow.com/questions/263400/what-is-the-best-algorithm-for-an-overridden-system-object-gethashcode/263416#263416
                int hash = 17;
                hash = hash * 23 + CenterX.GetHashCode();
                hash = hash * 23 + CenterY.GetHashCode();
                hash = hash * 23 + Width.GetHashCode();
                hash = hash * 23 + Height.GetHashCode();
                hash = hash * 23 + Angle.GetHashCode();

                return hash;
            }
        }

        public bool IsEmpty
        {
            get { return Width == 0 && Height == 0; }
        }

        public RectangleF ToRectangleF()
        {
            return new RectangleF(CenterX - Width / 2, CenterY - Height / 2, Width, Height);
        }

        public override string ToString()
        {
            var ci = System.Globalization.CultureInfo.InvariantCulture;
            return string.Format("{{ CenterX={0}, CenterY={1}, Width={2}, Height={3}, Angle={4} }}",
                CenterX.ToString("r", ci), CenterY.ToString("r", ci),
                Width.ToString("r", ci), Height.ToString("r", ci),
                Angle.ToString("r", ci));
        }

        public RotatedRectangleF Clone()
        {
            return new RotatedRectangleF(CenterX, CenterY, Width, Height, Angle);
        }

        public bool IntersectsWith(RotatedRectangleF rect)
        {
            var thisRect = Clone();
            var otherRect = rect.Clone();

            if (Angle != 0)
            {
                var angle = thisRect.Angle;
                thisRect.Angle = 0;

                var center = otherRect.Center;
                center.RotateAt(-angle, thisRect.Center);
                otherRect.Center = center;
                otherRect.Angle -= angle;
            }

            var thisEdges = new PointF[] { thisRect.GetLeftTopEdge(), thisRect.GetRigthTopEdge(), thisRect.GetRightBottomEdge(), thisRect.GetLeftBottomEdge() };
            var otherEdges = new PointF[] { otherRect.GetLeftTopEdge(), otherRect.GetRigthTopEdge(), otherRect.GetRightBottomEdge(), otherRect.GetLeftBottomEdge() };

            var firstProjectionIntersection = CheckProjectionIntersection(thisEdges, otherEdges);

            var secondProjectionIntersection = true;
            if (firstProjectionIntersection && Angle != rect.Angle)
            {
                thisRect = Clone();
                otherRect = rect.Clone();

                var angle = otherRect.Angle;
                otherRect.Angle = 0;

                var center = thisRect.Center;
                center.RotateAt(-angle, otherRect.Center);
                thisRect.Center = center;
                thisRect.Angle -= angle;

                thisEdges = new PointF[] { thisRect.GetLeftTopEdge(), thisRect.GetRigthTopEdge(), thisRect.GetRightBottomEdge(), thisRect.GetLeftBottomEdge() };
                otherEdges = new PointF[] { otherRect.GetLeftTopEdge(), otherRect.GetRigthTopEdge(), otherRect.GetRightBottomEdge(), otherRect.GetLeftBottomEdge() };

                secondProjectionIntersection = CheckProjectionIntersection(thisEdges, otherEdges);
            }

            return firstProjectionIntersection && secondProjectionIntersection;
        }

        private bool CheckProjectionIntersection(PointF[] points0, PointF[] points1)
        {
            float x00, x01, y00, y01;
            GetProjections(points0, out x00, out x01, out y00, out y01);

            float x10, x11, y10, y11;
            GetProjections(points1, out x10, out x11, out y10, out y11);

            return x00 <= x11 && x01 >= x10 && y00 <= y11 && y01 >= y10;
        }

        private void GetProjections(PointF[] points, out float x0, out float x1, out float y0, out float y1)
        {
            x0 = x1 = points[0].X;
            y0 = y1 = points[0].Y;

            for (int i = 1; i < points.Length; i++)
            {
                x0 = SMath.Min(x0, points[i].X);
                x1 = SMath.Max(x1, points[i].X);
                y0 = SMath.Min(y0, points[i].Y);
                y1 = SMath.Max(y1, points[i].Y);
            }
        }

        private PointF GetLeftTopEdge()
        {
            var p = new PointF(-Width / 2, -Height / 2);
            p.Rotate(Angle);
            p.Translate(CenterX, CenterY);
            return p;
        }

        private PointF GetRigthTopEdge()
        {
            var p = new PointF(Width / 2, -Height / 2);
            p.Rotate(Angle);
            p.Translate(CenterX, CenterY);
            return p;
        }

        private PointF GetRightBottomEdge()
        {
            var p = new PointF(Width / 2, Height / 2);
            p.Rotate(Angle);
            p.Translate(CenterX, CenterY);
            return p;
        }

        private PointF GetLeftBottomEdge()
        {
            var p = new PointF(-Width / 2, Height / 2);
            p.Rotate(Angle);
            p.Translate(CenterX, CenterY);
            return p;
        }
    }
}

File: AjaxVectorObjects/ClientScripts/Math/RotatedRectangleF.js
// Copyright (c) 2018 Aurigma Inc. All rights reserved.
// Licensed under the MIT License. See LICENSE file in the project root for full license information.
//
Type.registerNamespace("Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math");

Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF = function (centerX, centerY, width, height, angle) {
    ///	<summary>This client-side class corresponds to the <see cref="T:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF" /> server-side class.</summary>
    /// <remarks><para>This class represents a set of floating-point numbers that define the location, size, rotation angle and center of a rectangle and exposes a number of methods to operate with it.</para></remarks>
    /// <seealso cref="T:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF" />
    /// <param name="centerX" type="Number" />
    /// <param name="centerY" type="Number" />
    /// <param name="width" type="Number" />
    /// <param name="height" type="Number" />
    /// <param name="angle" type="Number" />
    /// <field name="CenterX" type="Number" integer="true" static="true"><summary>The x-coordinate of the rotation center of this rectangle.</summary></field>
    /// <field name="CenterY" type="Number" integer="true" static="true"><summary>The y-coordinate of the rotation center of this rectangle.</summary></field>
    /// <field name="Width" type="Number" integer="true" static="true"><summary>The width of this rectangle.</summary></field>
    /// <field name="Height" type="Number" integer="true" static="true"><summary>The height of this rectangle.</summary></field>
    /// <field name="Angle" type="Number" integer="true" static="true"><summary>The rotation angle of this rectangle.</summary></field>
    /// <constructor>
    /// 	<summary>Creates and initializes an instance of the <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RectangleF" /> class.</summary>
    /// 	<param name="centerX">The x-coordinate of the rotation center of the rectangle to create.</param>
    /// 	<param name="centerY">The y-coordinate of the rotation center of the rectangle to create.</param>
    /// 	<param name="width">The width of the rectangle to create.</param>
    /// 	<param name="height">The height of the rectangle to create.</param>
    /// 	<param name="angle">The rotation angle of the rectangle to create.</param>
    /// </constructor>
    Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF.initializeBase(this);

    this.CenterX = (centerX) ? centerX : 0;
    this.CenterY = (centerY) ? centerY : 0;
    this.Width = (width || width === 0) ? width : 2;
    this.Height = (height || height === 0) ? height : 2;
    this.Angle = (angle) ? angle : 0;
}

Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF.prototype = {
    clone: function () {
        /// <summary>Creates a new object that is a copy of the current <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF" /> instance.</summary>
        /// <returns type="Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF">A new <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF" /> that is a copy of this instance.</returns>
        return new Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF(this.CenterX, this.CenterY, this.Width, this.Height, this.Angle);
    },

    get_bounds: function () {
        /// <summary>Gets the bounds of this <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF" /> taking into account its rotation.</summary>
        /// <value type="Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RectangleF">The rectangle circumscribed about this <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF" />.</value>
        /// <remarks><para>This property corresponds to <see cref="P:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF.Bounds">RotatedRectangleF.Bounds</see> server-side member.</para></remarks>
        var m = Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math;
        var center = new m.PointF(this.CenterX, this.CenterY);
        var w = this.Width;
        var h = this.Height;
        var points = [
			new m.PointF(-w / 2, -h / 2),
			new m.PointF(-w / 2, h / 2),
			new m.PointF(w / 2, -h / 2),
			new m.PointF(w / 2, h / 2)
        ];
        var minx = 0, miny = 0, maxx = 0, maxy = 0;
        for (var i = 0; i < 4; i++) {
            points[i].rotate(this.Angle);
            if (points[i].X < minx || i === 0)
                minx = points[i].X;
            if (points[i].Y < miny || i === 0)
                miny = points[i].Y;
            if (points[i].X > maxx || i === 0)
                maxx = points[i].X;
            if (points[i].Y > maxy || i === 0)
                maxy = points[i].Y;
        }
        return new m.RectangleF(minx + center.X, miny + center.Y, maxx - minx, maxy - miny);
    },

    isEqual: function (rect, tolerance) {
        /// <summary>Determines whether the specified <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF" /> is equal to the current <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF" />.</summary>
        /// <param name="rect" type="Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF">The rotated rectangle to compare with the current <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF" />.</param>
        /// <returns type="Boolean"><strong>true</strong> if <paramref name="rect" /> is equal to the current <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF" />; otherwise, <strong>false</strong>.</returns>

        var m = Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math;
        return rect != null && m.EqualsOfFloatNumbers(this.CenterX, rect.CenterX, tolerance) && m.EqualsOfFloatNumbers(this.CenterY, rect.CenterY, tolerance) &&
            m.EqualsOfFloatNumbers(this.Width, rect.Width, tolerance) && m.EqualsOfFloatNumbers(this.Height, rect.Height, tolerance) &&
            m.EqualsOfFloatNumbers(this.Angle, rect.Angle, tolerance);
    },

    _getUpperLeftCorner: function () {
        var p = new Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF(-this.Width / 2, -this.Height / 2);
        p = p.rotate(this.Angle);
        p = p.translate(this.CenterX, this.CenterY);
        return p;
    },

    _getUpperRightCorner: function () {
        var p = new Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF(this.Width / 2, -this.Height / 2);
        p = p.rotate(this.Angle);
        p = p.translate(this.CenterX, this.CenterY);
        return p;
    },

    _getBottomRightCorner: function () {
        var p = new Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF(this.Width / 2, this.Height / 2);
        p = p.rotate(this.Angle);
        p = p.translate(this.CenterX, this.CenterY);
        return p;
    },

    _getBottomLeftCorner: function () {
        var p = new Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF(-this.Width / 2, this.Height / 2);
        p = p.rotate(this.Angle);
        p = p.translate(this.CenterX, this.CenterY);
        return p;
    },

    _checkProjectionIntersection: function (points0, points1) {
        var p0 = this._getProjections(points0);
        var p1 = this._getProjections(points1);

        return p0.left <= p1.right && p0.right >= p1.left && p0.top <= p1.bottom && p0.bottom >= p1.top;
    },

    _getProjections: function (points) {
        var x0 = x1 = points[0].X;
        var y0 = y1 = points[0].Y;

        for (var i = 1; i < points.length; i++) {
            x0 = Math.min(x0, points[i].X);
            x1 = Math.max(x1, points[i].X);
            y0 = Math.min(y0, points[i].Y);
            y1 = Math.max(y1, points[i].Y);
        }

        return { left: x0, right: x1, top: y0, bottom: y1 };
    },

    intersectsWith: function (rect) {
        var thisRect = this.clone();
        var otherRect = rect.clone();

        if (this.Angle != 0) {
            var angle = thisRect.Angle;
            thisRect.Angle = 0;

            var center = otherRect.get_center();
            center.rotateAt(-angle, thisRect.get_center());
            otherRect.set_center(center);
            otherRect.Angle -= angle;
        }

        var thisEdges = [thisRect._getUpperLeftCorner(), thisRect._getUpperRightCorner(), thisRect._getBottomRightCorner(), thisRect._getBottomLeftCorner()];
        var otherEdges = [otherRect._getUpperLeftCorner(), otherRect._getUpperRightCorner(), otherRect._getBottomRightCorner(), otherRect._getBottomLeftCorner()];

        var firstProjectionIntersection = this._checkProjectionIntersection(thisEdges, otherEdges);

        var secondProjectionIntersection = true;
        if (firstProjectionIntersection && this.Angle != rect.Angle) {
            thisRect = this.clone();
            otherRect = rect.clone();

            var angle = otherRect.Angle;
            otherRect.Angle = 0;

            var center = thisRect.get_center();
            center.rotateAt(-angle, otherRect.get_center());
            thisRect.set_center(center);
            thisRect.Angle -= angle;

            thisEdges = [thisRect._getUpperLeftCorner(), thisRect._getUpperRightCorner(), thisRect._getBottomRightCorner(), thisRect._getBottomLeftCorner()];
            otherEdges = [otherRect._getUpperLeftCorner(), otherRect._getUpperRightCorner(), otherRect._getBottomRightCorner(), otherRect._getBottomLeftCorner()];

            secondProjectionIntersection = this._checkProjectionIntersection(thisEdges, otherEdges);
        }

        return firstProjectionIntersection && secondProjectionIntersection;
    },

    get_location: function () {
        /// <summary>Gets or sets the coordinates of the upper-left corner of this <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF" />.</summary>
        /// <value type="Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF">The value which represents the upper-left corner of this <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF" />.</value>
        /// <remarks><para>This property corresponds to <see cref="P:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF.Location">RotatedRectangleF.Location</see> server-side member.</para></remarks>
        return this._getUpperLeftCorner();
    },

    set_location: function (location) {
        var p = this._getUpperLeftCorner();
        this.CenterX = this.CenterX + (location.X - p.X);
        this.CenterY = this.CenterY + (location.Y - p.Y);
    },

    get_center: function () {
        return new Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF(this.CenterX, this.CenterY);
    },

    set_center: function (center) {
        this.CenterX = center.X;
        this.CenterY = center.Y;
    },

    rotateAt: function (angle, center) {
        var rectCenter = this.get_center();
        rectCenter.rotateAt(angle, center);
        this.set_center(rectCenter);
        this.Angle += angle;
    },

    set_transform: function (transform) {
        if (transform == null || transform.isEqual(new Aurigma.GraphicsMill.AjaxControls.VectorObjects.Transform()))
            return;

        this.Width *= transform.get_scaleX();
        this.Height *= transform.get_scaleY();
        this.Angle += transform.get_angle();
        this.CenterX += transform.get_translateX();
        this.CenterY += transform.get_translateY();
    },

    get_transform: function (rectangle) {
        if (rectangle == null || this.isEqual(rectangle))
            return new Aurigma.GraphicsMill.AjaxControls.VectorObjects.Transform();

        var scaleX = rectangle.Width === 0 ? 0 : this.Width / rectangle.Width;
        var scaleY = rectangle.Height === 0 ? 0 : this.Height / rectangle.Height;
        var angle = this.Angle - rectangle.Angle;
        var center = rectangle.get_center();
        var translateX = this.CenterX - center.X;
        var translateY = this.CenterY - center.Y;

        return new Aurigma.GraphicsMill.AjaxControls.VectorObjects.Transform(scaleX, scaleY, translateX, translateY, angle);
    },

    toString: function () {
        var r = [this.CenterX, this.CenterY, this.Width, this.Height, this.Angle];
        return r.join(',');
    },

    toRectangleF: function () {
        return new Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RectangleF(this.CenterX - this.Width / 2, this.CenterY - this.Height / 2, this.Width, this.Height);
    },

    CenterX: 0,
    CenterY: 0,
    Width: 2,
    Height: 2,
    Angle: 0
}

Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF.fromRectangleF = function (rectangle) {
    return new Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF(
        rectangle.Left + rectangle.Width / 2, rectangle.Top + rectangle.Height / 2, rectangle.Width, rectangle.Height, 0);
};

Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF.FromLTRB = function (left, top, right, bottom) {
    return Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF.fromRectangleF(
        Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RectangleF.FromLTRB(left, top, right, bottom));
};

Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF.registerClass("Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.RotatedRectangleF");